cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -1828,6 +1828,148 @@ void AddUtf8Length(FunctionRegistry* registry) {
+            if (strings->length() > 0) {
+              KERNEL_RETURN_IF_ERROR(ctx, builder.Append(strings->GetView(0)));
+              for (int64_t j = 1; j < strings->length(); j++) {
+                KERNEL_RETURN_IF_ERROR(ctx, builder.AppendCurrent(separator));
+                KERNEL_RETURN_IF_ERROR(ctx, builder.AppendCurrent(strings->GetView(j)));
+              }
+            } else {
+              KERNEL_RETURN_IF_ERROR(ctx, builder.AppendEmptyValue());
+            }
+          }
+        }
+      } else if (batch[1].kind() == Datum::SCALAR) {
+        const auto& separator_scalar =
+            checked_cast<const BaseBinaryScalar&>(*batch[1].scalar());
+        if (!separator_scalar.is_valid) {
+          KERNEL_ASSIGN_OR_RAISE(
+              auto nulls, ctx,
+              MakeArrayOfNull(list.value_type(), list.length(), ctx->memory_pool()));
+          *output = *nulls->data();
+          output->type = list.value_type();
+          return;
+        }
+        util::string_view separator(*separator_scalar.value);
+
+        for (int64_t i = 0; i < list.length(); ++i) {
+          const std::shared_ptr<Array> slice = list.value_slice(i);
+          const ArrayType* strings = static_cast<const ArrayType*>(slice.get());
+          if ((strings->null_count() > 0) || (list.IsNull(i))) {
+            KERNEL_RETURN_IF_ERROR(ctx, builder.AppendNull());
+          } else {
+            if (strings->length() > 0) {
+              KERNEL_RETURN_IF_ERROR(ctx, builder.Append(strings->GetView(0)));
+              for (int64_t j = 1; j < strings->length(); j++) {
+                KERNEL_RETURN_IF_ERROR(ctx, builder.AppendCurrent(separator));
+                KERNEL_RETURN_IF_ERROR(ctx, builder.AppendCurrent(strings->GetView(j)));
+              }
+            } else {
+              KERNEL_RETURN_IF_ERROR(ctx, builder.AppendEmptyValue());
+            }
+          }
+        }
+      }
+      std::shared_ptr<Array> string_array;
+
+      KERNEL_RETURN_IF_ERROR(ctx, builder.Finish(&string_array));
+      *output = *string_array->data();
+      // correct the output type based on the input
+      output->type = list.value_type();
+    }
+  }
+};
+
+const FunctionDoc binary_join_doc(
+    "Join list of strings together with a `seperator` to form a single string",
+    ("Insert `seperator` between each list element, and concatenate them."),
+    {"list", "separator"});
+
+void AddJoin(FunctionRegistry* registry) {
+  auto func =
+      std::make_shared<ScalarFunction>("binary_join", Arity::Binary(), &binary_join_doc);
+  for (const std::shared_ptr<DataType>& ty : BaseBinaryTypes()) {
+    auto exec = GenerateTypeAgnosticVarBinaryBase<Join>(*ty);
+    DCHECK_OK(
+        func->AddKernel({InputType::Array(list(ty)), InputType::Scalar(ty)}, ty, exec));
+    DCHECK_OK(
+        func->AddKernel({InputType::Array(list(ty)), InputType::Array(ty)}, ty, exec));
+    // do we want to support scalar[list[str]] with array[str] ?
```

Comment:
It doesn't sound very useful. Either way, if it's not supported, then simply remove the commented out code.